packages/grafana-ui/src/themes/ThemeContext.tsx

```diff
@@ -39,9 +41,23 @@ export function useTheme(): GrafanaTheme {
   return useContext(ThemeContextMock || ThemeContext);
 }
 
-/** Hook for using memoized styles with access to the theme. */
+/**
+ * Hook for using memoized styles with access to the theme.
+ *
+ * NOTE: For memoization to work, you need to ensure that the function
+ * you pass in doesn't change, or only if it needs to. (i.e. declare
+ * your style creator outside of a function component or use `useCallback()`.)
+ * */
 export function useStyles<T>(getStyles: (theme: GrafanaTheme) => T) {
-  return stylesFactory(getStyles)(useTheme());
+  const theme = useTheme();
+
+  let memoizedStyleCreator = memoizedStyleCreators.get(getStyles);
+  if (!memoizedStyleCreator) {
+    memoizedStyleCreator = stylesFactory(getStyles);
+    memoizedStyleCreators.set(getStyles, memoizedStyleCreator);
+  }
+
+  return memoizedStyleCreator(theme);
```

Comment:
This looks good @s-h-a-d-o-w . WDYT about removing the memorized styles getter from the `memoizedStyleCreators` Map, when the component unmounts?

Reply:
I think that's a great catch and something I should really work out because it would constitute a memory leak otherwise. ;)

Reply:
Done.